types/node/index.d.ts

```diff
@@ -510,7 +510,7 @@ declare namespace NodeJS {
     type ExitListener = (code: number) => void;
     type RejectionHandledListener = (promise: Promise<any>) => void;
     type UncaughtExceptionListener = (error: Error) => void;
-    type UnhandledRejectionListener = (reason: any, promise: Promise<any>) => void;
+    type UnhandledRejectionListener = (reason: Error | {} | undefined, promise: Promise<any>) => void;
```

Comment:
Are you sure about this?
If I execute `Promise.reject([ 1, 2 ]);` the reason is `[ 1, 2 ]`.

Reply:
The empty object type (`{}`) refers to the type in your example. You can think of the empty object type as similar to `any` but, unlike `any`, it doesn't disable type checking. The user may validate and narrow the type with user-defined type guards.

More information about `{}` vs `any` here:
- https://github.com/Microsoft/TypeScript/issues/9999#issuecomment-238009695
- https://github.com/Microsoft/TypeScript/issues/17440

Hopefully one day we'll have a true `unknown` type: https://github.com/Microsoft/TypeScript/issues/10715

Reply:
As an example, currently the `any` type allows this without any errors:

``` ts
(reason: any) => {
  reason.foo.bar // no error :-(
}
```

By changing `any` to `{}`, we do get an error. The user must then validate and narrow the type in order to use it. This is of course much more type safe:

``` ts
const checkIsObjectWithFoo = (value: {}): value is { foo: { bar: {} } } => 'foo' in value;

(reason: {}) => {
  reason.foo.bar // error :-)
}

(reason: {}) => {
  checkIsObjectWithFoo(reason).foo.bar // no error :-)
}
```

Reply:
That's nice, thanks!
I think `| null` is needed as `Promise.reject(null);` is possible and `null` seems to not match `{}`.

Reply:
> By changing any to {}, we do get an error.

And that's the problem. There are probably some project relying on the current behavior and would break after this change.

Is it intended to break consumers in a patch release? 

If you want that extra type safety, you can always annotate the parameter of your callback.

Reply:
Perhaps we can hold off making this change until the new major release? I don't know how those get coordinated, mind you. Is there one planned/upcoming, do you know?

Reply:
Hi @ajafff, let me know what you think to my previous comment!

Reply:
My learning here is to never ever add `any` in typings as there is no way to get rid of it without the risk to break something. Versions of typings are dependent on versioning of node therefore no chance to reflect this in the version.